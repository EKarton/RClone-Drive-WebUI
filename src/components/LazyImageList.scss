.imagelist {
  height: 100%;

  &__row {
    display: flex;
    padding-bottom: 0.25rem;
  }
}
