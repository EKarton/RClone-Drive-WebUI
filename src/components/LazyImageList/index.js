import PropTypes from 'prop-types';
import AutoSizer from 'react-virtualized-auto-sizer';
import { FixedSizeList } from 'react-window';
import LazyImage from 'components/LazyImage';
import './index.scss';

export default function LazyImageList({ images, remote, onImageClicked, ...otherProps }) {
  const parseImageInfo = (fileName) => {
    const year = parseInt(fileName.substring(0, 4));
    const month = parseInt(fileName.substring(4, 6));
    const day = parseInt(fileName.substring(6, 8));

    return { year, month, day };
  };

  const parsedImages = images
    .map((item) => {
      const filePath = item.Path;
      const fileName = item.Name;
      const dirPath = filePath.substring(0, filePath.lastIndexOf('/'));
      const dateTaken = parseImageInfo(fileName);

      return { remote, dirPath, fileName, dateTaken };
    })
    .sort((image1, image2) => image2.fileName.localeCompare(image1.fileName));

  const handleImageClicked = (image) => () => {
    onImageClicked(image);
  };

  const renderCell =
    (width, height, numImagesPerRow) =>
    ({ index, style }) => {
      const selectedImages = [];

      let i = 0;
      while (i < numImagesPerRow && numImagesPerRow * index + i < parsedImages.length) {
        selectedImages.push(parsedImages[numImagesPerRow * index + i]);
        i += 1;
      }

      return (
        <div className="imagelist__row" style={style}>
          {selectedImages.map((selectedImage) => (
            <div
              key={selectedImage.fileName}
              onClick={handleImageClicked(selectedImage)}
              data-testid={selectedImage.fileName}
            >
              <LazyImage image={selectedImage} width={width} height={height} />
            </div>
          ))}
        </div>
      );
    };

  return (
    <AutoSizer {...otherProps}>
      {({ height, width }) => {
        const numImagesPerRow = width < 1920 ? 3 : 5;
        const imgWidth = Math.floor(width / numImagesPerRow);
        const imgHeight = imgWidth;

        const numRows = Math.ceil(images.length / numImagesPerRow);

        return (
          <FixedSizeList
            width={width}
            height={height}
            itemCount={numRows}
            itemSize={imgHeight}
          >
            {renderCell(imgWidth, imgHeight, numImagesPerRow)}
          </FixedSizeList>
        );
      }}
    </AutoSizer>
  );
}

LazyImageList.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      Path: PropTypes.string,
      Name: PropTypes.string,
    })
  ),
  remote: PropTypes.shape,
  onImageClicked: PropTypes.func,
};
