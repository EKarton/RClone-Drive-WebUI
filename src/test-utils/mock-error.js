/**
 * Replaces the error object's stack trace with mock data
 * @param {Error} error the Error object
 * @returns {Error} the same error, but with the stack trace mocked
 */
export function mockErrorStackTrace(error) {
  error.stack = 'Mock content for stack trace';
  return error;
}
