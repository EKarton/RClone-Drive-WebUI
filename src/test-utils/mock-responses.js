export const mockFiles = {
  list: [
    {
      Path: '.rclone',
      Name: '.rclone',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-11-01T21:09:05.237Z',
      IsDir: true,
      ID: '1a4mI4vsGKypqFFPwLnrFid1ZdvlhirDi',
    },
    {
      Path: 'Documents',
      Name: 'Documents',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T01:51:05.982Z',
      IsDir: true,
      ID: '1_za5Kqy9ZjXZ7rsztGpKhtJLTuNCKv8q',
    },
    {
      Path: 'Games',
      Name: 'Games',
      Size: -9,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T05:14:59.237Z',
      IsDir: true,
      ID: '1j6mAjxQoJokiU8yWJYgyWYIiSIgH1G7M',
    },
    {
      Path: 'Literature',
      Name: 'Literature',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-29T07:10:05.059Z',
      IsDir: true,
      ID: '1dj1T5YkjRQZNiQ5efCA8PzideJb88d3K',
    },
    {
      Path: 'Pictures',
      Name: 'Pictures',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:39:11.850Z',
      IsDir: true,
      ID: '1D2PpmoIT23NqyTnolmohZ0UxvOd2D7n4',
    },
    {
      Path: 'Software',
      Name: 'Software',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:45:02.471Z',
      IsDir: true,
      ID: '1zzr0Bwp9tel66rCUKz8mZIXRBR2kuywr',
    },
    {
      Path: 'Videos',
      Name: 'Videos',
      Size: -8,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T04:22:24.997Z',
      IsDir: true,
      ID: '1eJlJ1PT7b9XY3Q5tLeU49BNkug259986',
    },
    {
      Path: 'backup.sh',
      Name: 'backup.sh',
      Size: 1110,
      MimeType: 'application/x-sh',
      ModTime: '2021-11-14T18:58:44.655Z',
      IsDir: false,
    },
  ],
};

export const mockRemotes = {
  remotes: ['googledrive', 'onedrive', 'icloud'],
};

export const mockPictures = {
  list: [
    {
      Path: 'Pictures/2012/20120517_171428.JPG',
      Name: '20120517_171428.JPG',
      Size: 335408,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:14:30.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_135206.JPG',
      Name: '20120517_135206.JPG',
      Size: 262719,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T17:52:06.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_204557.JPG',
      Name: '20120516_204557.JPG',
      Size: 245109,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T00:45:58.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_174510.JPG',
      Name: '20120517_174510.JPG',
      Size: 271914,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:45:12.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120820_144833.JPG',
      Name: '20120820_144833.JPG',
      Size: 2662880,
      MimeType: 'image/jpeg',
      ModTime: '2012-08-20T18:48:32.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_094409.JPG',
      Name: '20120517_094409.JPG',
      Size: 280596,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T13:44:10.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181954.JPG',
      Name: '20120516_181954.JPG',
      Size: 416424,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:19:54.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181809.JPG',
      Name: '20120516_181809.JPG',
      Size: 211575,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:18:10.000Z',
      IsDir: false,
    },
  ],
};
