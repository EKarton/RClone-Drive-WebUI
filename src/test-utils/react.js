import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
