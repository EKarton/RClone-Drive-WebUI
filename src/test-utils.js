import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
