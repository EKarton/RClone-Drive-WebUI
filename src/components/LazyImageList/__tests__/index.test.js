import { mockPictures } from 'test-utils/mock-responses';
import { customRender, userEvent, screen } from 'test-utils/react';
import LazyImageList from '..';

jest.mock('components/LazyImage', () => () => null);

describe('LazyImageList', () => {
  // Derived from https://github.com/bvaughn/react-virtualized/issues/493#issuecomment-447014986
  const originalOffsetHeight = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    'offsetHeight'
  );
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    'offsetWidth'
  );

  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
      configurable: true,
      value: 800,
    });
  });

  afterAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', originalOffsetHeight);
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth);
  });

  it.each([800, 2000])('should match snapshot given width is %s', (width) => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
      configurable: true,
      value: width,
    });

    const remote = 'gdrive';
    const onImageClicked = jest.fn();

    const { baseElement } = customRender(
      <LazyImageList
        images={mockPictures.list}
        remote={remote}
        onImageClicked={onImageClicked}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onImageClicked() correctly when clicked on an image', () => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
      configurable: true,
      value: 800,
    });

    const remote = 'gdrive';
    const onImageClicked = jest.fn();

    customRender(
      <LazyImageList
        images={mockPictures.list}
        remote={remote}
        onImageClicked={onImageClicked}
      />
    );

    userEvent.click(screen.getByTestId('20120517_171428.JPG'));

    expect(onImageClicked).toBeCalledWith({
      remote,
      dirPath: 'Pictures/2012',
      fileName: '20120517_171428.JPG',
      dateTaken: {
        year: 2012,
        month: 5,
        day: 17,
      },
    });
  });
});
