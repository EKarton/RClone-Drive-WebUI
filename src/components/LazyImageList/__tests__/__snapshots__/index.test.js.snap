// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImageList should match snapshot given width is 800 1`] = `
<body>
  <div>
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    >
      <div
        style="position: relative; height: 800px; width: 800px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 798px; width: 100%;"
        >
          <div
            class="imagelist__row"
            style="position: absolute; left: 0px; top: 0px; height: 266px; width: 100%;"
          >
            <div
              data-testid="20120820_144833.JPG"
            />
            <div
              data-testid="20120517_174510.JPG"
            />
            <div
              data-testid="20120517_171428.JPG"
            />
          </div>
          <div
            class="imagelist__row"
            style="position: absolute; left: 0px; top: 266px; height: 266px; width: 100%;"
          >
            <div
              data-testid="20120517_135206.JPG"
            />
            <div
              data-testid="20120517_094409.JPG"
            />
            <div
              data-testid="20120516_204557.JPG"
            />
          </div>
          <div
            class="imagelist__row"
            style="position: absolute; left: 0px; top: 532px; height: 266px; width: 100%;"
          >
            <div
              data-testid="20120516_181954.JPG"
            />
            <div
              data-testid="20120516_181809.JPG"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 801px; height: 801px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</body>
`;

exports[`LazyImageList should match snapshot given width is 2000 1`] = `
<body>
  <div>
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    >
      <div
        style="position: relative; height: 800px; width: 2000px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 800px; width: 100%;"
        >
          <div
            class="imagelist__row"
            style="position: absolute; left: 0px; top: 0px; height: 400px; width: 100%;"
          >
            <div
              data-testid="20120820_144833.JPG"
            />
            <div
              data-testid="20120517_174510.JPG"
            />
            <div
              data-testid="20120517_171428.JPG"
            />
            <div
              data-testid="20120517_135206.JPG"
            />
            <div
              data-testid="20120517_094409.JPG"
            />
          </div>
          <div
            class="imagelist__row"
            style="position: absolute; left: 0px; top: 400px; height: 400px; width: 100%;"
          >
            <div
              data-testid="20120516_204557.JPG"
            />
            <div
              data-testid="20120516_181954.JPG"
            />
            <div
              data-testid="20120516_181809.JPG"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 2001px; height: 801px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</body>
`;
